### PR TITLE
Modules and Webhooks Support

### PR DESCRIPTION
**Pull request by Hou-Rui**
Add support for the front-end to fetch modules from the back-end;
Add Webhooks support in the back-end.

### CODE REVIEW

**Comment by martin-henz** [comment]
This PR decreases test coverage by quite a bit. Can you try to maintain the coverage, by writing test cases?

**Comment by BoAi01** [comment]
Test cases added. Coverage can maintain 91% (0.09% lower).

File: lib/cadet_web/controllers/module_controller.ex [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    Synchronize local module files witn GitHub repository via WebHook
+    and repond HTTP resquest from frontend
+    """
+  
+    use CadetWeb, :controller
+  
+    use PhoenixSwagger
+  
+    def handle(conn, _payload) do
+      repo = Git.new("priv/module_static")
```
Comment (martin-henz) [inline lib/cadet_web/controllers/module_controller.ex:12]: wondering about the choice of name for the folder. This is the destination of the webhook, right? So priv/module_static gets populated by the modules from the modules repo, right? In that case, what is "static" about it?

File: priv/module_static/lib/binary_tree.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
+// tree.js: Binary Tree abstraction for M5
```
Comment (martin-henz) [inline priv/module_static/lib/binary_tree.js:1]: Having libraries explicitly included here will violate the "single source of truth" principle. We would have two versions and it wouldn't be clear what version is available in what situation.

File: priv/module_static/lib/runes.js [hunk 3]
```diff
@@ -0,0 +1,6463 @@
+(function (exports) {
```
Comment (martin-henz) [inline priv/module_static/lib/runes.js:1]: same as for binary_tree.js

File: lib/cadet_web/controllers/module_controller.ex (hunk 1 above)
Comment (BoAi01) [reply lib/cadet_web/controllers/module_controller.ex:12]: Okay, we can change the name of the destination folder, what about putting the module files under priv/modules?

File: priv/module_static/lib/binary_tree.js (hunk 2 above)
Comment (BoAi01) [reply priv/module_static/lib/binary_tree.js]: Okay, I have added the files under priv/modules to .gitignore.

File: priv/module_static/lib/runes.js (hunk 3 above)
Comment (BoAi01) [reply priv/module_static/lib/runes.js]: Same as above. Modules files are added to .gitignore.

File: priv/module_static/lib/runes.js [hunk 4]
```diff
@@ -6407,7 +6413,7 @@
   getReadyWebGLForCanvas('3d');
 
   exports.anaglyph = anaglyph;
-  exports.animate = animate;
+  exports.animated_show = animated_show;
```
Comment (travisryte) [inline priv/module_static/lib/runes.js]: How can I test this?

Comment (Hou-Rui) [reply priv/module_static/lib/runes.js]: You need a journey-frontend to test this. The whole new lib/runes requires to be run on the journey-frontend.

**Comment by angelsl** [comment]
Going to close this as we are thinking of a backend-independent module system.

**Pull request closed without merging**